src/mono/mono/mini/aot-compiler.c

```diff
@@ -6628,13 +6629,24 @@ emit_and_reloc_code (MonoAotCompile *acfg, MonoMethod *method, guint8 *code, gui
 					}
 				} else if (patch_info->type == MONO_PATCH_INFO_JIT_ICALL_ID) {
 					MonoJitICallInfo * const info = mono_find_jit_icall_info (patch_info->data.jit_icall_id);
-					const char * const sym = info->c_symbol;
-					if (!got_only && sym && acfg->aot_opts.direct_icalls && info->func == info->wrapper) {
-						/* Call to a jit icall without a wrapper */
-						direct_call = TRUE;
-						external_call = TRUE;
-						g_assert (strlen (sym) < 1000);
-						direct_call_target = g_strdup_printf ("%s%s", acfg->user_symbol_prefix, sym);
+					if (!got_only && info->func == info->wrapper) {
+						const char * sym = NULL;
+						if (patch_info->data.jit_icall_id == MONO_JIT_ICALL_mono_dummy_runtime_init_callback) {
+							g_assert (acfg->aot_opts.static_link && acfg->aot_opts.runtime_init_callback != NULL);
+							sym = acfg->aot_opts.runtime_init_callback;
+							direct_call = TRUE;
+							external_call = TRUE;
+						} else if (info->c_symbol && acfg->aot_opts.direct_icalls) {
+							sym = info->c_symbol;
+							direct_call = TRUE;
+							external_call = TRUE;
+						}
+
+						if (sym) {
+							/* Call to a jit icall without a wrapper */
+							g_assert (strlen (sym) < 1000);
+							direct_call_target = g_strdup_printf ("%s%s", acfg->user_symbol_prefix, sym);
```

Comment:
Why would we do this here and not in method-to-ir when we see a call to `MONO_JIT_ICALL_mono_dummy_runtime_init_callback` ?

Reply:
Because it must be a direct icall, and this is the place where we convert icalls into direct icalls. Normal icalls depends on GOT meaning runtime must be initialized when called, direct icall is just a straight call to a C symbol that doesn't need the runtime to be initialized and since this call is to the callback function doing lazy initialize of the runtime it must be a direct icall.

Reply:
I also don't think this is the right place to do this, it won't handle llvm for example.

Reply:
@vargaz  but we won't compile native-to-managed wrapper using LLVM anyways, correct? Since we need to covnert it to a direct icall, do we have a better place to do that lowering?

Reply:
We could also add change to mono_aot_get_direct_call_symbol, that is used in llvmonly mode and we could make emit_and_realloc call that function as well, then we would have the logic in one place.

Reply:
Looking at it, emitting these direct calls is a bit messy right now, so this might be the best location for it until this code gets cleaned up.

Reply:
Could one of you make an issue to clean up direct calls?

Reply:
I added logic to detect the direct icall in mono_aot_get_direct_call_symbol, that should cover the llvmonly use case as well. Regular LLVM backend should not compile the native-to-managed wrapper so should not be affected.

Created, https://github.com/dotnet/runtime/issues/82405 for future cleanup of direct icall logic.